hs/src/Reach/Eval/Core.hs

```diff
@@ -310,15 +321,45 @@ env_merge_ imode left righte = foldlM (env_insertp_ imode) left $ M.toList right
 env_merge :: HasCallStack => SLEnv -> SLEnv -> App SLEnv
 env_merge = env_merge_ DisallowShadowing
 
+-- Utilities to track variable
+
+useStrict :: App Bool
+useStrict = liftIO . readIORef =<< asks (sco_use_strict . e_sco)
+
+whenUsingStrict :: App () -> App ()
+whenUsingStrict e = useStrict >>= flip when e
+
+shouldNotTrackVariable :: (SrcLoc, SLVar) -> Bool
+shouldNotTrackVariable (SrcLoc _ _ (Just ReachStdLib), _) = True
+shouldNotTrackVariable (_, "main") = True
+shouldNotTrackVariable (_, "_") = True
+shouldNotTrackVariable _ = False
+
+trackVariable :: (SrcLoc, SLVar) -> App ()
+trackVariable el =
+  whenUsingStrict $ do
+    unused_vars <- asks e_unused_variables
+    unless (shouldNotTrackVariable el) $
+      liftIO $ modifyIORef unused_vars (el :)
+
+markVarUsed :: (SrcLoc, SLVar) -> App ()
+markVarUsed v =
+  whenUsingStrict $ do
+    unused_vars <- asks e_unused_variables
+    liftIO $ modifyIORef unused_vars $ filter (v /=)
```

Comment:
I think the data type would be more efficient if it were a set. You have this linear rather than logarithmic

Reply:
My `IORef` idea would be even more efficient, because the environment would just also store a `Maybe (IORef Bool)` and then just set it to false and then the unused thing would run the ones set to true.